### PR TITLE
Support graphql oneOf directive without composeDirective

### PR DESCRIPTION
**Pull request by jdolle**
Apollo doesn't support oneOf because they are using an out dated version of graphql.

This PR adds functionality to keep the oneOf directives through composition and adds a test to ensure this works.

This avoids having to unnecessarily add `@composeDirective` for `@oneOf`. `@composeDirective` should be reserved for custom directives instead of native directives.

### CODE REVIEW

File: __tests__/composition.spec.ts [hunk 1]
```diff
@@ -8647,4 +8647,56 @@ testImplementations((api) => {
       }
     `);
   });
+
+  test("guild composition supports composing built-in directive (@oneOf)", () => {
+    api.runIf("guild", () => {
```
Comment (jdolle) [inline __tests__/composition.spec.ts]: only run for guild b/c others dont support atm

File: src/subgraph/state.ts [hunk 2]
```diff
@@ -2252,6 +2262,15 @@ function inputObjectTypeFactory(state: SubgraphState) {
     setTag(typeName: string, tag: string) {
       getOrCreateInputObjectType(state, typeName).tags.add(tag);
     },
+    setOneOf(
+      typeName: string,
+    ) {
+      getOrCreateInputObjectType(state, typeName).isOneOf =
+        true;
+      const oneOfDirective = getOrCreateDirective(state, 'oneOf');
```
Comment (jdolle) [inline src/subgraph/state.ts]: must ensure the definition is added or else `Supergraph.build` fails because the internal version of graphql it's using is out of date

**Comment by kamilkisiela** [comment]
Few things:
- What will happen if subgraph A already has `@oneOf` defined via `@composeDirective`, and after we introduce native support for it, subgraph B adds `@oneOf` without `@composeDirective`
- Let's add a few scenarios where you try to "break" it, like two subgraphs defining the same input type, a case where the input type is defined twice in two subgraphs, but fields differ, where one defines `@oneOf` and the other does not etc

**Comment by jdolle** [comment]
I've added tests and adjusted the logic to be more explicit to address your concerns around these edge cases

File: src/subgraph/state.ts [hunk 3]
```diff
@@ -2252,6 +2262,35 @@ function inputObjectTypeFactory(state: SubgraphState) {
     setTag(typeName: string, tag: string) {
       getOrCreateInputObjectType(state, typeName).tags.add(tag);
     },
+    setOneOf(typeName: string) {
+      const existingOneOf = state.types.get("oneOf");
+
+      // if definition matches the build-in, flag the type as isOneOf
```
Comment (jdolle) [inline src/subgraph/state.ts]:
```suggestion
      // if definition matches the built-in, flag the type as isOneOf
```

**Review by kamilkisiela — APPROVED** [review]
